SpringDelegatingLambdaContainerHandler seems does not work with query parameters

Hello to everyone.
Spring boot 3.1.0
aws-serverless-java-container-springboot3:2.0.0-M2
Frontend service: ALB, API_GW, Functional URL
Deployment method: terraform or manual

## Scenario
I want my java lambda microservice to process all types of events without additional expenses: ALB, API_GW, HTTP, Functional URL

## Expected behavior
Behave the same way to all event types listed above

## Actual behavior
When I use SpringDelegatingLambdaContainerHandler, I get error 400 every time I have query parameters for ALB event. if I use functional url event, I get internal server error.

## Steps to reproduce
`@RequestMapping(path = "/w6/test/{uuid}/refresh", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<TMaxmindResponse> refresh(
            @Parameter(description = "UUID of the order") @PathVariable("uuid") String uuid,
            @Parameter(description = "ID of the order(primary key)") @RequestParam("order_id") Long orderId,
            @Parameter(description = "tab") @RequestParam("tab") String tab
    ) `



## Full log output
```
jakarta.servlet.ServletException: Request processing failed: java.lang.NullPointerException: Cannot read the array length because \"\u003cparameter1\u003e\" is null\n\tat org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1019)\n\tat org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)\n\tat jakarta.servlet.http.HttpServlet.service(HttpServlet.java:527)\n\tat org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)\n\tat jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc$ProxyFilterChain$ServletFilterProxy.doFilter(ProxyMvc.java:280)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc$ProxyFilterChain.doFilter(ProxyMvc.java:233)\n\tat org.springframework.web.filter.AbstractRequestLoggingFilter.doFilterInternal(AbstractRequestLoggingFilter.java:289)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc$ProxyFilterChain.doFilter(ProxyMvc.java:233)\n\tat org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc$ProxyFilterChain.doFilter(ProxyMvc.java:233)\n\tat org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:109)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc$ProxyFilterChain.doFilter(ProxyMvc.java:233)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc.service(ProxyMvc.java:144)\n\tat org.springframework.cloud.function.serverless.web.ProxyMvc.service(ProxyMvc.java:138)\n\tat com.amazonaws.serverless.proxy.spring.SpringDelegatingLambdaContainerHandler.handleRequest(SpringDelegatingLambdaContainerHandler.java:74)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\n\tat com.amazonaws.services.lambda.runtime.api.client.EventHandlerLoader$StreamMethodRequestHandler.handleRequest(EventHandlerLoader.java:378)\n\tat com.amazonaws.services.lambda.runtime.api.client.EventHandlerLoader$2.call(EventHandlerLoader.java:905)\n\tat com.amazonaws.services.lambda.runtime.api.client.AWSLambda.startRuntime(AWSLambda.java:245)\n\tat com.amazonaws.services.lambda.runtime.api.client.AWSLambda.startRuntime(AWSLambda.java:197)\n\tat com.amazonaws.services.lambda.runtime.api.client.AWSLambda.main(AWSLambda.java:187)\nCaused by: java.lang.NullPointerException: Cannot read the array length because \"\u003cparameter1\u003e\" is null\n\tat java.base/java.io.ByteArrayInputStream.\u003cinit\u003e(Unknown Source)\n\tat org.springframework.cloud.function.serverless.web.ProxyHttpServletRequest.getInputStream(ProxyHttpServletRequest.java:280)\n\tat org.springframework.web.util.ContentCachingRequestWrapper.getInputStream(ContentCachingRequestWrapper.java:98)\n\tat org.springframework.http.server.ServletServerHttpRequest.getBody(ServletServerHttpRequest.java:206)\n\tat org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodArgumentResolver$EmptyBodyCheckingHttpInputMessage.\u003cinit\u003e(AbstractMessageConverterMethodArgumentResolver.java:323)\n\tat org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodArgumentResolver.readWithMessageConverters(AbstractMessageConverterMethodArgumentResolver.java:172)\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor.readWithMessageConverters(RequestResponseBodyMethodProcessor.java:163)\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor.resolveArgument(RequestResponseBodyMethodProcessor.java:136)\n\tat org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:122)\n\tat org.springframework.web.method.support.InvocableHandlerMethod.getMethodArgumentValues(InvocableHandlerMethod.java:181)\n\tat org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:148)\n\tat org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:884)\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:797)\n\tat org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)\n\tat org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1081)\n\tat org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:974)\n\tat org.springframework.web.servlet.FrameworkServlet.processReque
```

I tried to wrap my head around it for 2 or 3 days, and found that when it tries to do 

`mvc.service(httpServletRequest, httpServletResponse);`

it fails to error 400, no matter which parameters are. Also, if method parameters marked as optional, it returns exception and error 500, for instance, `@RequestParam("order_id") Optional<Long> orderId,`

After spending that much time, I decided that this is a bug and I really appreciate if this bug is resolved ASAP. Thanks

--- a/aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java
+++ b/aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java
@@ -2,7 +2,10 @@ package com.amazonaws.serverless.proxy.spring;
 
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
 import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
@@ -10,6 +13,7 @@ import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.springframework.cloud.function.serverless.web.ServerlessHttpServletRequest;
 import org.springframework.cloud.function.serverless.web.ServerlessMVC;
+import org.springframework.util.CollectionUtils;
 import org.springframework.util.FileCopyUtils;
 import org.springframework.util.MultiValueMapAdapter;
 import org.springframework.util.StringUtils;
@@ -109,6 +113,9 @@ class AwsSpringHttpProcessingUtils {
 		AwsProxyRequest v1Request = readValue(request, AwsProxyRequest.class, mapper);
 		
 		ServerlessHttpServletRequest httpRequest = new ServerlessHttpServletRequest(servletContext, v1Request.getHttpMethod(), v1Request.getPath());
+
+		populateQueryStringparameters(v1Request.getQueryStringParameters(), httpRequest);
+		
 		if (v1Request.getMultiValueHeaders() != null) {
 			MultiValueMapAdapter headers = new MultiValueMapAdapter(v1Request.getMultiValueHeaders());
 			httpRequest.setHeaders(headers);
@@ -128,16 +135,21 @@ class AwsSpringHttpProcessingUtils {
 				securityWriter.writeSecurityContext(v1Request, lambdaContext));
 		return httpRequest;
 	}
+	
+	
 
 	@SuppressWarnings({ "rawtypes", "unchecked" })
 	private static HttpServletRequest generateRequest2(String request, Context lambdaContext,
 			SecurityContextWriter securityWriter, ObjectMapper mapper, ServletContext servletContext) {
 		HttpApiV2ProxyRequest v2Request = readValue(request, HttpApiV2ProxyRequest.class, mapper);
+		
+		
 		ServerlessHttpServletRequest httpRequest = new ServerlessHttpServletRequest(servletContext,
 				v2Request.getRequestContext().getHttp().getMethod(), v2Request.getRequestContext().getHttp().getPath());
+		populateQueryStringparameters(v2Request.getQueryStringParameters(), httpRequest);
 		
 		v2Request.getHeaders().forEach(httpRequest::setHeader);
-	
+		
 		if (StringUtils.hasText(v2Request.getBody())) {
 			httpRequest.setContentType("application/json");
 			httpRequest.setContent(v2Request.getBody().getBytes(StandardCharsets.UTF_8));
@@ -149,6 +161,14 @@ class AwsSpringHttpProcessingUtils {
 		httpRequest.setAttribute(RequestReader.JAX_SECURITY_CONTEXT_PROPERTY,
 				securityWriter.writeSecurityContext(v2Request, lambdaContext));
 		return httpRequest;
+	}
+	
+	private static void populateQueryStringparameters(Map<String, String> requestParameters, ServerlessHttpServletRequest httpRequest) {
+		if (!CollectionUtils.isEmpty(requestParameters)) {
+			for (Entry<String, String> entry : requestParameters.entrySet()) {
+				httpRequest.setParameter(entry.getKey(), entry.getValue());
+			}
+		}
 	}
 	
 	private static <T> T readValue(String json, Class<T> clazz, ObjectMapper mapper) {

--- a/aws-serverless-java-container-springboot3/src/test/java/com/amazonaws/serverless/proxy/spring/SpringDelegatingLambdaContainerHandlerTests.java
+++ b/aws-serverless-java-container-springboot3/src/test/java/com/amazonaws/serverless/proxy/spring/SpringDelegatingLambdaContainerHandlerTests.java
@@ -13,6 +13,7 @@ import java.util.Map;
 
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.MethodSource;
+import org.springframework.cloud.function.serverless.web.ServerlessServletContext;
 import org.springframework.util.CollectionUtils;
 
 import com.amazonaws.serverless.proxy.spring.servletapp.MessageData;
@@ -72,11 +73,15 @@ public class SpringDelegatingLambdaContainerHandlerTests {
             + "    },\n"
             + "    \"queryStringParameters\": {\n"
             + "        \"abc\": \"xyz\",\n"
+            + "        \"name\": \"Ricky\",\n"
             + "        \"foo\": \"baz\"\n"
             + "    },\n"
             + "    \"multiValueQueryStringParameters\": {\n"
             + "        \"abc\": [\n"
             + "            \"xyz\"\n"
+            + "        ],\n"
+            + "        \"name\": [\n"
+            + "            \"Ricky\"\n"
             + "        ],\n"
             + "        \"foo\": [\n"
             + "            \"bar\",\n"
@@ -124,7 +129,7 @@ public class SpringDelegatingLambdaContainerHandlerTests {
             "  \"version\": \"2.0\",\n" +
             "  \"routeKey\": \"$default\",\n" +
             "  \"rawPath\": \"/my/path\",\n" +
-            "  \"rawQueryString\": \"parameter1=value1&parameter1=value2&parameter2=value\",\n" +
+            "  \"rawQueryString\": \"parameter1=value1&parameter1=value2&name=Ricky&parameter2=value\",\n" +
             "  \"cookies\": [\n" +
             "    \"cookie1\",\n" +
             "    \"cookie2\"\n" +
@@ -135,6 +140,7 @@ public class SpringDelegatingLambdaContainerHandlerTests {
             "  },\n" +
             "  \"queryStringParameters\": {\n" +
             "    \"parameter1\": \"value1,value2\",\n" +
+            "    \"name\": \"Ricky\",\n" +
             "    \"parameter2\": \"value\"\n" +
             "  },\n" +
             "  \"requestContext\": {\n" +
@@ -202,6 +208,22 @@ public class SpringDelegatingLambdaContainerHandlerTests {
         return Arrays.asList(new String[]{API_GATEWAY_EVENT, API_GATEWAY_EVENT_V2});
     }
 
+    @MethodSource("data")
+    @ParameterizedTest
+    public void validateComplesrequest(String jsonEvent) throws Exception {
+        initServletAppTest();
+        InputStream targetStream = new ByteArrayInputStream(this.generateHttpRequest(jsonEvent, "POST", 
+        		"/foo/male/list/24", "{\"name\":\"bob\"}", null));
+        ByteArrayOutputStream output = new ByteArrayOutputStream();
+        handler.handleRequest(targetStream, output, null);
+        Map result = mapper.readValue(output.toString(StandardCharsets.UTF_8), Map.class);
+        assertEquals(200, result.get("statusCode"));
+        String[] respponseBody = ((String) result.get("body")).split("/");
+        assertEquals("male", respponseBody[0]);
+        assertEquals("24", respponseBody[1]);
+        assertEquals("Ricky", respponseBody[2]);
+    }
+    
     @MethodSource("data")
     @ParameterizedTest
     public void testAsyncPost(String jsonEvent) throws Exception {

--- a/aws-serverless-java-container-springboot3/src/test/java/com/amazonaws/serverless/proxy/spring/servletapp/ServletApplication.java
+++ b/aws-serverless-java-container-springboot3/src/test/java/com/amazonaws/serverless/proxy/spring/servletapp/ServletApplication.java
@@ -1,9 +1,13 @@
 package com.amazonaws.serverless.proxy.spring.servletapp;
 
 import org.springframework.boot.autoconfigure.SpringBootApplication;
-import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
-import org.springframework.context.annotation.ComponentScan;
 import org.springframework.context.annotation.Import;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
 
 @SpringBootApplication(exclude = {
         org.springframework.boot.autoconfigure.security.reactive.ReactiveUserDetailsServiceAutoConfiguration.class,
@@ -14,5 +18,15 @@ import org.springframework.context.annotation.Import;
         org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration.class
 })
 @Import(MessageController.class)
+@RestController
 public class ServletApplication {
+	
+	@RequestMapping(path = "/foo/{gender}/list/{age}", method = RequestMethod.POST, produces = MediaType.APPLICATION_JSON_VALUE)
+    public String complexRequest(
+            @PathVariable("gender") String gender,
+            @PathVariable("age") String age,
+            @RequestParam("name") String name
+    ) {
+		return gender + "/" + age + "/" + name;
+	}
 }

--- a/samples/springboot3/alt-pet-store/README.md
+++ b/samples/springboot3/alt-pet-store/README.md
@@ -37,3 +37,20 @@ PetStoreApi - URL for application            https://xxxxxxxxxx.execute-api.us-w
 
 $ curl https://xxxxxxxxxx.execute-api.us-west-2.amazonaws.com/pets
 ```
+
+You can also try a complex request passing both path and request parameters to  complex endpoint such as this:
+
+
+```
+@RequestMapping(path = "/foo/{gender}/bar/{age}", method = RequestMethod.POST, produces = MediaType.APPLICATION_JSON_VALUE)
+public String complexRequest(@RequestBody String body,
+        @PathVariable("gender") String gender,
+        @PathVariable("age") String age,
+        @RequestParam("name") String name
+)
+```
+For example.
+
+```
+curl -d '{"key1":"value1", "key2":"value2"}' -H "Content-Type: application/json" -X POST https://zuhd709386.execute-api.us-east-2.amazonaws.com/foo/male/bar/25?name=Ricky
+```

--- a/samples/springboot3/alt-pet-store/src/main/java/com/amazonaws/serverless/sample/springboot3/controller/PetsController.java
+++ b/samples/springboot3/alt-pet-store/src/main/java/com/amazonaws/serverless/sample/springboot3/controller/PetsController.java
@@ -17,6 +17,8 @@ package com.amazonaws.serverless.sample.springboot3.controller;
 import com.amazonaws.serverless.sample.springboot3.model.Pet;
 import com.amazonaws.serverless.sample.springboot3.model.PetData;
 
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
@@ -32,6 +34,7 @@ import java.util.UUID;
 @RestController
 @EnableWebMvc
 public class PetsController {
+	
     @RequestMapping(path = "/pets", method = RequestMethod.POST)
     public Pet createPet(@RequestBody Pet newPet) {
         if (newPet.getName() == null || newPet.getBreed() == null) {
@@ -73,5 +76,15 @@ public class PetsController {
         newPet.setName(PetData.getRandomName());
         return newPet;
     }
+    
+    @RequestMapping(path = "/foo/{gender}/bar/{age}", method = RequestMethod.POST, produces = MediaType.APPLICATION_JSON_VALUE)
+    public String complexRequest(@RequestBody String body,
+            @PathVariable("gender") String gender,
+            @PathVariable("age") String age,
+            @RequestParam("name") String name
+    ) {
+    	System.out.println("Body: " + body + " - " + gender + "/" + age + "/" + name);
+		return gender + "/" + age + "/" + name;
+	}
 
 }

